In [18]:
# Importando Bibliotecas

import numpy as np
from matplotlib import pyplot as plt
import pandas as pd
import seaborn as sns
from IPython.display import display, Markdown
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error, r2_score

%matplotlib inline

In [2]:
# Importando o Dataset
 
from sklearn.datasets import load_boston
boston_dataset = load_boston()

C:\Users\gabri\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (

In [3]:
# Obtendo as chaves da dict 'boston_dataset'

boston_dataset.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])

In [28]:
# Carregando o dataset com o Pandas

display(Markdown("**Dataset:**"))
boston_x = pd.DataFrame(data=boston_dataset.data, columns=boston_dataset.feature_names)
boston_x

**Dataset:**

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [27]:
# Carregando a variável target do dataset com o Pandas

display(Markdown("**Variável Target:**"))
boston_y = pd.DataFrame(data=boston_dataset.target, columns=['MEDV'])
boston_y

**Variável Target:**

MEDV
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
..    ...
501  22.4
502  20.6
503  23.9
504  22.0
505  11.9

[506 rows x 1 columns]

In [26]:
# Exibindo as estatísticas descritivas do dataset

display(Markdown("**Estatísticas descritivas do Dataset:**"))
boston_x.describe()

**Estatísticas descritivas do Dataset:**

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  506.000000  
mean    12.653063  
std      7.141062  
min      1.730000  
25%      6.950000  
50%     11.360000  
75%     16.955000  
max     37.970000

In [29]:
# Preparando a normalização min-max (coletando características min-max das features)

scaler = MinMaxScaler().fit(boston_x)
scaler

MinMaxScaler()

In [30]:
# Aplicando a normalização min-max no dataset sem a variável target

boston_x = scaler.transform(boston_x)
boston_x

array([[0.00000000e+00, 1.80000000e-01, 6.78152493e-02, ...,
        2.87234043e-01, 1.00000000e+00, 8.96799117e-02],
       [2.35922539e-04, 0.00000000e+00, 2.42302053e-01, ...,
        5.53191489e-01, 1.00000000e+00, 2.04470199e-01],
       [2.35697744e-04, 0.00000000e+00, 2.42302053e-01, ...,
        5.53191489e-01, 9.89737254e-01, 6.34657837e-02],
       ...,
       [6.11892474e-04, 0.00000000e+00, 4.20454545e-01, ...,
        8.93617021e-01, 1.00000000e+00, 1.07891832e-01],
       [1.16072990e-03, 0.00000000e+00, 4.20454545e-01, ...,
        8.93617021e-01, 9.91300620e-01, 1.31070640e-01],
       [4.61841693e-04, 0.00000000e+00, 4.20454545e-01, ...,
        8.93617021e-01, 1.00000000e+00, 1.69701987e-01]])

In [31]:
# Divisão do dataset em conjunto de treinamento e conjunto de teste (para variáveis independentes e variável target)

x_train, x_test, y_train, y_test = train_test_split(boston_x, boston_y, test_size=0.2, random_state=5)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(404, 13)
(102, 13)
(404, 1)
(102, 1)


In [32]:
# Transformando o array das variáveis independentes em tensores

tensor_x_train = torch.tensor(x_train, dtype=torch.float32)
tensor_x_test = torch.tensor(x_test, dtype=torch.float32)
print(tensor_x_train.shape)
print(tensor_x_test.shape)

torch.Size([404, 13])
torch.Size([102, 13])


In [33]:
# Transformando o array da variável target de treinamento em tensor

tensor_y_train = torch.tensor(y_train.values, dtype=torch.float32)
tensor_y_train.shape

torch.Size([404, 1])

In [34]:
# Declarando hiperparâmetros da rede neural (número de entradas, número de saídas, número de neurônios nas camadas ocultas e taxa de aprendizado)

n_input, n_output, n_hidden, learning_rate = tensor_x_train.size()[1], 1, 30, 0.01

In [35]:
# Construindo a rede neural com duas camadas ocutas lineares

model = nn.Sequential(
    nn.Linear(n_input, n_hidden),
    nn.Linear(n_hidden, n_output)
)
model

Sequential(
  (0): Linear(in_features=13, out_features=30, bias=True)
  (1): Linear(in_features=30, out_features=1, bias=True)
)

In [37]:
# Definindo o otimizador que será utilizado no backpropagation

loss_function = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

In [38]:
# Treinando a rede neural com forward propagation e backpropagation

losses = []
for epoch in range(1000):
    pred_y = model(tensor_x_train)
    loss = loss_function(pred_y, tensor_y_train)
    losses.append(loss.item())
    
    model.zero_grad()
    loss.backward()
    
    optimizer.step()



In [39]:
# Avaliando o desempenho do modelo construido

y_train_predict = model(tensor_x_train).detach().numpy()

rmse = np.sqrt(mean_squared_error(y_train_predict, y_train))
r2 = r2_score(y_train, y_train_predict)

display(Markdown("**Performance do modelo para dados de treinamento**"))
print('RMSE: {}'.format(rmse))
print('R2: {}'.format(r2))
print("\n")

print("--------------------------------------")

y_test_predict = model(tensor_x_test).detach().numpy()

rmse = np.sqrt(mean_squared_error(y_test_predict, y_test))
r2 = r2_score(y_test, y_test_predict)

display(Markdown("**Performance do modelo para dados de teste**"))
print('RMSE: {}'.format(rmse))
print('R2: {}'.format(r2))
print("\n")


**Performance do modelo para dados de treinamento**

RMSE: 4.808082965341093
R2: 0.7308823637189505


--------------------------------------


**Performance do modelo para dados de teste**

RMSE: 4.527104127175611
R2: 0.738234013500048




- Pode-se observar que o modelo explica 73.1% das amostras de treinamento e 73.8% das amostras de teste
- Não há indícios de sobreajuste